File: app/src/main/java/com/dumuzeyn/mp3player/PlaybackStatisticsPolicy.java
package com.dumuzeyn.mp3player;

final class PlaybackStatisticsPolicy {
    static final long LONG_TRACK_THRESHOLD_MS = 30_000L;
    static final long SKIP_THRESHOLD_MS = 10_000L;

    private PlaybackStatisticsPolicy() {
    }

    static boolean countsAsPlay(long listenedMs, long durationMs) {
        if (listenedMs <= 0L) {
            return false;
        }
        long required = durationMs > 0L && durationMs < 60_000L
                ? Math.max(5_000L, durationMs / 2L)
                : LONG_TRACK_THRESHOLD_MS;
        return listenedMs >= required;
    }

    static boolean countsAsSkip(long listenedMs, long durationMs, boolean userInitiated) {
        return userInitiated && listenedMs >= 0L && listenedMs < SKIP_THRESHOLD_MS
                && (durationMs <= 0L || listenedMs < durationMs / 2L);
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/PlaybackStatisticsPolicy.kt
package com.dumuzeyn.mp3player

object PlaybackStatisticsPolicy {
    const val LONG_TRACK_THRESHOLD_MS = 30_000L
    const val SKIP_THRESHOLD_MS = 10_000L

    @JvmStatic
    fun countsAsPlay(listenedMs: Long, durationMs: Long): Boolean {
        if (listenedMs <= 0L) return false
        val required = if (durationMs in 1L..<60_000L) {
            maxOf(5_000L, durationMs / 2L)
        } else {
            LONG_TRACK_THRESHOLD_MS
        }
        return listenedMs >= required
    }

    @JvmStatic
    fun countsAsSkip(listenedMs: Long, durationMs: Long, userInitiated: Boolean): Boolean =
        userInitiated &&
            listenedMs in 0L..<SKIP_THRESHOLD_MS &&
            (durationMs <= 0L || listenedMs < durationMs / 2L)
}
